### PR TITLE
Add DHT key-value behaviour

### PR DESCRIPTION
**Pull request by montekki**
Drag `KademliaOut` events corresponding to DHT record system all the way to `NetworkSpecialization` to be usable in polkadot's code. Requests for records get/put are forwarded through `NetworkService` down to `Kademlia` via `get_value` and `put_value` methods.

### CODE REVIEW

**Comment by tomaka** [comment]
Why/how can putting a key in the DHT fail?
Also, when that happens we should just try again instead of simply reporting the failure.

File: core/network/src/protocol/event.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
```
Comment (bkchr) [inline core/network/src/protocol/event.rs]:
```suggestion
// Copyright 2019 Parity Technologies (UK) Ltd.
```

File: core/network/src/behaviour.rs [hunk 2]
```diff
@@ -14,12 +14,14 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
 
-use crate::{debug_info, discovery::DiscoveryBehaviour, discovery::DiscoveryOut, DiscoveryNetBehaviour};
+use crate::{debug_info, discovery::DiscoveryBehaviour, discovery::DiscoveryOut,
+	DiscoveryNetBehaviour, event::DhtEvent};
```
Comment (bkchr) [inline core/network/src/behaviour.rs]:
```suggestion
	debug_info, discovery::DiscoveryBehaviour, discovery::DiscoveryOut, DiscoveryNetBehaviour, event::DhtEvent
};
```

File: core/network/src/behaviour.rs [hunk 3]
```diff
@@ -14,12 +14,14 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
 
-use crate::{debug_info, discovery::DiscoveryBehaviour, discovery::DiscoveryOut, DiscoveryNetBehaviour};
+use crate::{debug_info, discovery::DiscoveryBehaviour, discovery::DiscoveryOut,
```
Comment (bkchr) [inline core/network/src/behaviour.rs]:
```suggestion
use crate::{
```

**Comment by montekki** [comment]
@tomaka so, what could happen in the kademlia level is:

1. The upper code requests to put a `key` into the `dht`.
2. Kademlia looks into it's buckets and asks it's peers to find the `K` closest `PeerId`s to the key that should be the holders.
3. Kademlia starts `PUT_VALUE` requests to all of the `K` found peers.
4. All those requests fail in any way: timeout, host is unreachable, etc.

Now with regards to behavior on failures: I believe that the ideal solution would be to to follow the logic of the `libp2p` in a way that we make our implementation compatible with others and let the kademlia itself handle re-inserting, updates etc. But that is a bit not the task currently at hand.

What is at hand is the behaviour in polkadot: to me seems that it should be responsible for handling `PutValueFailed` cases: by the time we understand that we have failed to add a value the the DHT, we might be needing to put another sessionkey or something. But that's all just my speculatios.

File: core/network/src/discovery.rs [hunk 4]
```diff
@@ -81,12 +82,35 @@ impl<TSubstream> DiscoveryBehaviour<TSubstream> {
...
 /// Event generated by the `DiscoveryBehaviour`.
 pub enum DiscoveryOut {
 	/// We have discovered a node. Can be called multiple times with the same identity.
 	Discovered(PeerId),
+
+	/// The DHT yeided results for the record request, grouped in (key, value) pairs.
+	ValueFound(Vec<(Multihash, Vec<u8>)>),
+
+	/// The record requested was not found in the DHT.
+	ValueNotFound, // TODO: key of the value requested
```
Comment (rphmeier) [inline core/network/src/discovery.rs]: TODOs need issues filed.

**Comment by rphmeier** [comment]
Generally looks OK to me, I'll wait for a sign-off by someone more aware of network code (@tomaka or @gterzian )

File: core/network/src/discovery.rs (hunk 4 above)
Comment (tomaka) [inline core/network/src/discovery.rs]: This TODO can be fixed now that #2945 has been merged.

File: core/network/src/behaviour.rs [hunk 5]
```diff
@@ -112,6 +121,16 @@ impl<TBehaviour, TBehaviourEv, TSubstream> Behaviour<TBehaviour, TBehaviourEv, T
 	pub fn user_protocol_mut(&mut self) -> &mut TBehaviour {
 		&mut self.user_protocol.0
 	}
+
+	/// Get a record from the DHT.
+	pub fn get_value(&mut self, key: &Multihash) {
+		self.discovery.get_value(key);
+	}
+
+	/// Put a record into DHT.
```
Comment (tomaka) [inline core/network/src/behaviour.rs]:
```suggestion
	/// Starts putting a record into DHT. Will later produce either a `ValuePut` or a `ValuePutFailed` event.
```

**Pull request merged**